File: 155_min_stack.cpp
#include <vector>

class MinStack
{
public:
    MinStack()
    {
    }

    void push(int val)
    {
        _data.push_back(val);
        if (val < _min)
        {
            _min = val;
        }
    }

    void pop()
    {
        _data.pop_back();
    }

    int top()
    {
        return _data.back();
    }

    int getMin()
    {
        return _min;
    }

private:
    int _min = 0;
    std::vector<int> _data;
};

int main()
{
    MinStack *minStack = new MinStack();
    minStack->push(-2);
    minStack->push(0);
    minStack->push(-3);
    minStack->getMin(); // return -3
    minStack->pop();
    minStack->top();    // return 0
    minStack->getMin(); // return -2

    return 0;
}

File: ts/src/155_min_stack.ts
class MinStack {
    constructor() {

    }

    push(val: number): void {
        this.stack.push(val);

        if (this.stack.length == 1) {
            this.mins.push(val);
            return;
        }

        const currMin: number = this.mins[this.mins.length - 1];
        if (currMin > val) {
            this.mins.push(val);
        }
        else {
            this.mins.push(currMin);
        }

    }

    pop(): void {
        this.stack.pop();
        this.mins.pop();
    }

    top(): number {
        return this.stack[this.stack.length - 1];
    }

    getMin(): number {
        return this.mins[this.mins.length - 1];
    }

    stack: number[] = [];
    mins: number[] = [];
}

{
    let minStack: MinStack = new MinStack();
    minStack.push(-2);
    minStack.push(0);
    minStack.push(-3);
    minStack.getMin(); // return -3
    minStack.pop();
    minStack.top();    // return 0
    minStack.getMin(); // return -2
}